Update/Add functionalities for import storage domain

maorlipchuk opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Update register template module
Add register for VM
Add iSCSI login and iSCSI discover
Try to fix a typeError issue when adding storage domain
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
ovirt_hosts: Add iscsidiscover to ovirt_hosts

Adding functionality of iscsidiscover to be used to discover iscsi
targets.

Commit:
ovirt_templates: Update the argument spec of templates.

Add id of template since it is needed for register.

Commit:
ovirt_vms: Register unregistered VM.

Use register of VM with id instead of name since an
unregitered entity can be registered also without name attribute.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -426,7 +428,9 @@ def main():
         ),
+        id=dict(default=None),
         name=dict(required=True),
+        import_=dict(type='bool', default=False),
```
We use an `imported` state for importing a template, maybe we should use it also here. Also I think that other then `iscsi` and `fcp` storage can be imported,no?

maorlipchuk replied:
sure, I will try to indicate a new state for that
all storage domains can be imported basically, but it works differently for NFS and block.
in NFS if the storage exists then the engine assumes we mean to import it, in block we need to specifically indicate the flag

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -217,7 +216,8 @@ def main():
         ),
-        name=dict(default=None, required=True),
+        id=dict(default=None),
+        name=dict(default=None),
```
can you please add `required_one_of=[['id', 'name']]`

maorlipchuk replied:
sure, will do

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -324,34 +324,34 @@ def main():
 
             # Find the the unregistered Template we want to register:
-            templates = templates_service.list(unregistered=True)
+            templates = templates_service.list(query={'unregistered': 'true'})
```
What's not working here? The previous code, should be fine.

maorlipchuk replied:
I will take a look at it again, and will fix or add a comment

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -324,34 +324,34 @@ def main():
+            templates = templates_service.list(query={'unregistered': 'true'})
             template = next(
-                (t for t in templates if t.name == module.params['name']),
+                (t for t in templates if t.id == module.params['id']),
```
We should support both.

maorlipchuk replied:
yes, I will rephrase the condition

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -324,34 +324,34 @@ def main():
-                # Test if template is registered:
-                template = templates_module.search_entity()
-                if template is None:
+                template_service = templates_service.template_service(module.params['id'])
```
We should support both, `name` and `id`

maorlipchuk replied:
Will rephrase the condition

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -324,34 +324,34 @@ def main():
-                if module.params['wait']:
-                    template = wait_for_import(module, templates_service)
 
+                # Register template is merely a DB operation so there is no reason to wait.
```
Hmm, I think it will be better to wait, even if it's just DB operation, so you are sure, that when the task is done, you can work with template in another task.

maorlipchuk replied:
yes, there are no tasks, so when the engine command is ended we are sure that the Template is in the DB (or not if encountered an error), I will re-verify that this is the behavior again just to be sure.
Besides that, regarding the search of the entity, I encountered an issue here that the search for the entity is filtered with name, and since the entity has only id, I think that the search function should be changed

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -217,7 +216,8 @@ def main():
             default='present',
         ),
-        name=dict(default=None, required=True),
+        id=dict(default=None),
```
The `id` paramater should be documented.

maorlipchuk replied:
I will add an example, thanks

machacekondra replied:
I mean also in `DOCUMENTATION` constant not only in `EXAMPLES`.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -113,7 +113,6 @@
 # Register template
 - ovirt_templates:
   state: registered
-  name: mytemplate
```
please either add `id` as example, or leave the `name`

maorlipchuk replied:
sure, I will add a new example

machacekondra reviewed (commented):
This was fixed in latest Python SDK, please update it and this commit is not needed.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -336,6 +341,7 @@ def main():
         override_display=dict(default=None),
         kernel_params=dict(default=None, type='list'),
         hosted_engine=dict(default=None, choices=['deploy', 'undeploy']),
+        iscsi=dict(default=None, type='dict'),
```
Can you rename `iscsi` to `iscsi_address`, I think it's not needed to have it a dict, no? String should be fine.

Also please add

```python
required_if=[
  ['state', 'iscsidiscover', ['iscsi_address']]
]
```

And also please document this parameter.

maorlipchuk replied:
but we might also need to use username and password

maorlipchuk replied:
The request should be as follow:
- ovirt_hosts:
    state: iscsidiscover
    name: myhost
    iscsi:
         username: iscsi_user
         password: *****
         address: 10.34.61.145
How do you want to rename it?

machacekondra replied:
Ok, then this is fine I think. Thanks.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -386,6 +392,13 @@ def main():
+                # This is a synchronize action and we only want to
+                # know if we got 200 response code
+                action='iscsi_discover',
+                iscsi=otypes.IscsiDetails(address=module.params['iscsi']['address'])
```
Should this be enough? :

iscsi=otypes.IscsiDetails(address=module.params['iscsi_address'])

Also what will happen if that action was already invoked, do we get error, or HTTP 200?
The best behavior would be to not sent `iscsi_dicover` if it was already sent. And sent it only when needed, when it wasn't yet sent.

maorlipchuk replied:
> Should this be enough? 

  nope, I should also add it username and password
>  Also what will happen if that action was already invoked, do we get error, or HTTP 200?

  If the discover was already invoke we will get HTTP 200, same as will be done through the REST.

> The best behavior would be to not sent iscsi_dicover if it was already sent. And sent it only when needed, when it wasn't yet sent.

  But what if the user added targets to its storage server, and he would still want to call discover to discover the new targets.

maorlipchuk replied:
Regarding login, if we re-login I think that this should be handled through the engine, what do you think?

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -153,6 +153,11 @@
+# discover iscsi targets
+- ovirt_hosts:
+    state: iscsidiscover
+    name: myhost
```
Missing `iscsi_address` param here?

maorlipchuk replied:
will fix

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -42,7 +42,7 @@
         choices: [
             'present', 'absent', 'maintenance', 'upgraded', 'started',
-            'restarted', 'stopped', 'reinstalled', 'iscsidiscover'
+            'restarted', 'stopped', 'reinstalled', 'iscsidiscover', 'iscsilogin'
```
Same as `iscsidiscover`.

machacekondra replied:
Please add a note that `iscsilogin` is supported since 2.4. Something like:

```yaml
state:
         description:
             - "State which should a host to be in after successful completion."
             - "I(iscsilogin) and I(iscsidiscover) are supported since version 2.4."
```

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -354,6 +354,12 @@
+# Register VM
+ovirt_vms:
+    state: registered
+    storage_domain: mystorage
```
I think you miss `id` or `name` here.

maorlipchuk replied:
will fix

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1243,6 +1250,48 @@ def main():
+                try:
+                    vm = vm_service.get()
+                # TODO: TypeError is being thrown from _check_content_type, although it should be SDK.error
+                except TypeError:
```
It shouldn't be in new Python SDK.

maorlipchuk replied:
will remove

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1243,6 +1250,48 @@ def main():
+            vms_service = storage_domain_service.vms_service()
+
+            # Find the the unregistered VM we want to register:
+            vms = vms_service.list(query={'unregistered': 'true'})
```
Shouldn't this be OK?:

```python
vms = vms_service.list(unregistered=True)
```

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1243,6 +1250,48 @@ def main():
+
+            # Find the the unregistered VM we want to register:
+            vms = vms_service.list(query={'unregistered': 'true'})
+            vm = next(
```
This looks strange, I think you can just use:

In general:

1) if VM is already in system, you should just return that template, currently it's tryting to fetch that template from storage domain rather then system.

2) If VM is not in system, but it's in storage domain as unregistered, we should run `register` action. 

3) I think we should support both `id` and `name`.

maorlipchuk replied:
sure, I will keep the name logic

Commit:
ovirt_storage_domains: Add support for import block storage domain.

maorlipchuk commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -158,6 +158,11 @@
+- ovirt_hosts:
+    state: iscsilogin
+    name: myhost
+
```
This is an example of a call from a playbook:

                  - ovirt_hosts:
                          state: iscsilogin
                          # TODO: Choose active host in the DC
                          name: "{{ host_name }}"
                          auth: "{{ ovirt_auth }}"
                          iscsi:
                                  username: "{{ item.0.username }}"
                                  password: "{{ item.0.password }}"
                                  address: "{{ item.0.address }}"
                                  target: "{{ item.1 }}"
                    with_subelements:
                          - "{{ import_iSCSI }}"
                          - target

var file includes:
            - import_iSCSI:
                # TODO: Add support for connect to multiple targets with the same LUN.
        # Every connect should be done using a different ip
        - {id: "e796dbf3-0b60-4087-ac39-b67e05f062ee",
           address: "10.35.16.55",
           target: ["iqn.2015-07.com.mlipchuk2.redhat:444", "iqn.2015-07.com.mlipchuk3.redhat:444"],
           username: "",
           password: "",
           DC_name: "test",
           name: "iscsi"}

maorlipchuk commented:
Hi Ondra,
I verified those patches with the following playbook (see [1]).

[1]

                  - ovirt_storage_domains:
                          name: "initial_NFS"
                          # TODO: Choose active host in the DC
                          host: "{{ host_name }}"
                          data_center: "{{ item.DC_name }}"
                          auth: "{{ ovirt_auth }}"
                          nfs:
                                  address: "{{ item.address }}"
                                  path: "{{ item.path }}"
                    with_items:
                          - "{{ initial_NFS }}"

                  - ovirt_storage_domains:
                          # TODO: Choose active host in the DC
                          name: "{{ item.name }}"
                          host: "{{ host_name }}"
                          auth: "{{ ovirt_auth }}"
                          data_center: "{{ item.DC_name }}"
                          nfs:
                                  address: "{{ item.address }}"
                                  path: "{{ item.path }}"
                    with_items:
                          - "{{ import_NFS }}"

                            #                  - ovirt_hosts:
                            #                          state: iscsidiscover
                            #                          # TODO: Choose active host in the DC
                            #                          name: "{{ host_name }}"
                            #                          auth: "{{ ovirt_auth }}"
                            #                          iscsi:
                            #                                  username: "{{ item.username }}"
                            #                                  password: "{{ item.password }}"
                            #                                  address: "{{ item.address }}"
                            #                    with_items:
                            #                          - "{{ import_iSCSI }}"

                  - ovirt_hosts:
                          state: iscsilogin
                          # TODO: Choose active host in the DC
                          name: "{{ host_name }}"
                          auth: "{{ ovirt_auth }}"
                          iscsi:
                                  username: "{{ item.0.username }}"
                                  password: "{{ item.0.password }}"
                                  address: "{{ item.0.address }}"
                                  target: "{{ item.1 }}"
                    with_subelements:
                          - "{{ import_iSCSI }}"
                          - target

                  - ovirt_storage_domains:
                          state: 'import'
                          # TODO: Choose predefine host in the block DC
                          id: "{{ item.id }}"
                          name: "{{ item.name }}"
                          host: "{{ host_name }}"
                          auth: "{{ ovirt_auth }}"
                          data_center: "{{ item.DC_name }}"
                          # TODO: For import block there is no need for the iscsi parameters
                          iscsi:
                                  username: "{{ item.username }}"
                                  password: "{{ item.password }}"
                                  address: "{{ item.address }}"
                                  target: "{{ item.target[0] }}"
                    with_items:
                          - "{{ import_iSCSI }}"

                  - ovirt_templates:
                          state: registered
                          storage_domain: "{{ item.storage_domain }}"
                          auth: "{{ ovirt_auth }}"
                          id: "{{ item.id }}"
                          cluster: "{{ item.cluster }}"
                          # TODO: ignore_errors: yes
                    with_items:
                          - "{{ register_templates }}"

                  - ovirt_vms:
                          state: registered
                          storage_domain: "data6"
                          auth: "{{ ovirt_auth }}"
                          name: "VMTest1"
                          cluster: "cluster1"
                          # TODO: ignore_errors: yes

                  - ovirt_vms:
                          state: registered
                          storage_domain: "{{ item.storage_domain }}"
                          id: "{{ item.id }}"
                          auth: "{{ ovirt_auth }}"
                          cluster: "{{ item.cluster }}"
                          # TODO: ignore_errors: yes
                    with_items:
                          - "{{ register_vms }}"

Commit:
Add ovirt_storage_templates_facts

Adding fact module for storage templates.
The module should help with registering unregistered templates.

Commit:
Add ovirt_storage_vms_facts

Adding fact module for storage VMs.
The module should help with registering unregistered VMs.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -199,25 +206,38 @@ def post_import_action(self, entity):
+    if (module.params['id'] is not None):
+        # The template id can not be searched with templates_module.search_entity
+        # since the 'id' attribute is not searchable in the engine, therefore searching
+        # the id through templates_service.template_service.
```
Can we implement this in `ovirt.py` in `module_utils`?
Can you rewrite the `search_entity` to something like:

```python
   def search_entity(self, search_params=None):
        """
        Always first try to search by `ID`, if ID isn't specified,
        check if user constructed special search in `search_params`,
        if not search by `name`.
        """
        entity = None

        if self._module.params.get('id') is not None:
            entity = get_entity(self._service.service(self._module.params['id']))
        elif search_params is not None:
            entity = search_by_attributes(self._service, **search_params)
        elif self._module.params.get('name') is not None:
            entity = search_by_attributes(self._service, name=self._module.params['name'])

        return entity
```

maorlipchuk replied:
Yes, I thought about this also, but haven't managed to post a patch of it yet.
I will give it a try

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -386,6 +396,15 @@ def main():
                 fail_condition=failed_state,
             )
+        elif state == 'iscsidiscover':
+            ret = hosts_module.action(
```
We have an issue here. `hosts_module.action` will return the in `ret` the host details instead of `iscsi targets` details.
So please rewrite this to something like:

```python
host_id = get_id_by_name(module.params['name'])
iscsi_targets = hosts_service.service().iscsi_discover(
    iscsi=otypes.IscsiDetails(
        username=module.params['iscsi']['username'],
        password=module.params['iscsi']['password'],
        address=module.params['iscsi']['address'],
    ),
)
ret = {
    'changed': False,
    'id': host_id, 
    'iscsi_targets': iscsi_targets,
}
``` 

Then please also change the documentation of `RETURN` to something like:

```python
RETURN = '''
id:
    description: ID of the host which is managed
    returned: On success if host is found.
    type: str
    sample: 7de90f31-222c-436c-a1ca-7e655bd5b60c
host:
    description: "Dictionary of all the host attributes. Host attributes can be found on your oVirt/RHV instance
                  at following url: http://ovirt.github.io/ovirt-engine-api-model/master/#types/host."
    returned: On success if host is found.
    type: dict
iscsi_targets:
    description: "List of host iscsi targets"
    returned: On success if host is found and state is iscsi_discover.
    type: list
'''
```

Sorry that I didn't noticed it before.

maorlipchuk replied:
no problem, will do, thanks!

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1054,10 +1069,34 @@ def control_state(vm, vms_service, module):
+
+
+def _fetch_vm(module, vms_module, vms_service):
+    if (module.params['id'] is not None):
```
Same as in `ovirt_templates` patch, can you reimplent the support in `ovirt.py` ?

maorlipchuk replied:
will give it a try

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1054,10 +1069,34 @@ def control_state(vm, vms_service, module):
     argument_spec = ovirt_full_argument_spec(
         state=dict(
-            choices=['running', 'stopped', 'present', 'absent', 'suspended', 'next_run'],
+            choices=['running', 'stopped', 'present', 'absent', 'suspended', 'next_run', 'registered'],
```
Can you please add a note in documentation that `registered` is supported since 2.4?

maorlipchuk replied:
will do

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -423,9 +425,10 @@ def control_state(sd_module):
+            choices=['present', 'absent', 'maintenance', 'unattached', 'import'],
             default='present',
         ),
+        id=dict(default=None),
```
Please add doc for `id` attribute.

maorlipchuk replied:
will do

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -423,9 +425,10 @@ def control_state(sd_module):
     argument_spec = ovirt_full_argument_spec(
         state=dict(
-            choices=['present', 'absent', 'maintenance', 'unattached'],
+            choices=['present', 'absent', 'maintenance', 'unattached', 'import'],
```
We usually use past time, can you rename to 'imported'? and add a note it's supported since 2.4 in doc?

maorlipchuk replied:
yes, will do

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -465,8 +468,12 @@ def main():
-            sd_id = storage_domains_module.create()['id']
+        elif state == 'present' or state == 'import':
+            try:
+                sd = storage_domains_service.storage_domain_service(module.params['id']).get()
```
I believe this is not needed, when you change the code as I've suggested in the first patch.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2016 Red Hat, Inc.
```
2017

maorlipchuk replied:
oops, will change

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,131 @@
+DOCUMENTATION = '''
+---
+module: ovirt_storage_templates_facts
+short_description: Retrieve facts about one or more oVirt/RHV virtual machines relate to a storage domain.
```
s/virtual machines/templates ?

maorlipchuk replied:
yep, I will change this.
It is basically an WIP since I was not sure what is your opinion regarding new facts (Should I reuse the existing one or just add new one).
For now I added a new one

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,131 @@
+author: "Maor Lipchuk"
+version_added: "2.4"
+description:
+    - "Retrieve facts about one or more oVirt/RHV virtual machines relate to a storage domain."
```
s/virtual machines/templates ?

maorlipchuk replied:
done

machacekondra commented:
Thanks, I've posted few comments, if you can change it would be cool. Also can you please separate the last two commits with facts modules to separate PR?

maorlipchuk commented:
Thanks for the review, I'm on it

Commit:
Add functionality of partial import to unregistered VMs.

Commit:
Add functionality of partial import to unregistered Templates.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -386,6 +416,39 @@ def main():
+            }
+        elif state == 'iscsilogin':
+            host_id = get_id_by_name(hosts_service, module.params['name'])
+            iscsi_targets = hosts_service.service(host_id).iscsi_login(
```
`iscsi_login`[1] action don't return `iscsi_targets`, it don't return anything.

So in this case I think it's better to use:

```python
            ret = hosts_module.action(
                action='iscsi_login',
                iscsi=otypes.IscsiDetails(
                     port=module.params['iscsi']['port'] if module.params['iscsi']['port'] else None,
                     username=module.params['iscsi']['username'],
                     password=module.params['iscsi']['password'],
                     address=module.params['iscsi']['address'],
                     target=module.params['iscsi']['target'],
                 ),
            )
```

[1] https://github.com/oVirt/ovirt-engine-api-model/blob/master/src/main/java/services/HostService.java#L510

machacekondra replied:
Also `iscsi` param is required if using `iscsi_discover` or `iscsi_login` state, right?

So please add:

```python
required_if=[
  ['state', 'iscsi_login'', ['iscsi']],
  ['state', 'iscsi_discover'', ['iscsi']]
]
```

maorlipchuk replied:
will do, thanks

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -465,8 +473,13 @@ def main():
+                sd = search_entity()
+                sd_id = sd.id
+            except Exception:
+                sd_id = storage_domains_module.create()['id']
```
You don't need this code anymore, no?

Just this, should work fine:
```python
sd_id = storage_domains_module.create()['id']
```

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_domains.py:
```diff
@@ -256,6 +261,8 @@ def build_entity(self):
             name=self._module.params['name'],
             description=self._module.params['description'],
             comment=self._module.params['comment'],
+            import_=True if (self._module.params['state'] == 'import' and storage_type in ['iscsi', 'fcp']) else None,
```
Why only `iscsi` and `fcp` is supported?

maorlipchuk replied:
because I was trying to reflect the same behavior as being used also in the REST.
We use an "_import" boolean flag for block storage domains to tell whether it is a new storage domain or an imported one, because with FCP for example we don't mention any targets at all *that is why we added a new flag "_import")
I can use this flag as well for other storage domains if you think that it will be necessary

maorlipchuk replied:
IINM For NFS storage domain we do not mention any indication whether it is a new storage domain or an imported one. The engine simply check if there is an existing storage domain, and if there is it imports it

machacekondra replied:
Ok, thanks for explanation.

maorlipchuk replied:
sure, let me know if you think it will be better to add it to all type of storage domains

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,131 @@
+    - "This module creates a new top-level C(ovirt_storage_templates) fact, which
+       contains a list of templates."
+options:
+    pattern:
```
Well, the search is not supported for this service, so I would just add `unregistered` parameter, and remove the `pattern` as it's useless.

machacekondra replied:
Can you please change it? It don't make much sense to use it as is currently.

maorlipchuk replied:
sure, sorry, for some reason I missed that comment

maorlipchuk replied:
sure, sorry, I missed that for some reason

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_vms_facts.py:
```diff
@@ -0,0 +1,131 @@
+    - "This module creates a new top-level C(ovirt_storage_vms) fact, which
+       contains a list of virtual machines."
+options:
+    pattern:
```
Same as for templates.

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -40,9 +40,10 @@
     state:
```
You miss documentation for `iscsi` attribute. Please add it and please add there a note, that it's used only when state is `iscsidiscover` or `iscsilogin`. and don't forget to add `version_added: "2.4"`

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -113,9 +120,24 @@
+  cluster: mycluster
   name: mytemplate
+
+# Register template
```
Just please change the comment `Register template using id` or something similiar, so it's distinguished fromt the previsous example

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -113,9 +120,24 @@
+  cluster: mycluster
+  id: 1111-1111-1111-1111
+
+# Register template
```
Register template, allowing partial import

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -199,13 +221,13 @@ def post_import_action(self, entity):
         poll_interval = module.params['poll_interval']
         while time.time() < start + timeout:
-            template = search_by_name(templates_service, module.params['name'])
+            template = search_entity()
```
Shouldn't this be? 

```python
template = templates_module.search_entity()
```

maorlipchuk replied:
changed

maorlipchuk replied:
I guess that since the template registration eventually succeeded I missed that

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -263,6 +288,7 @@ def main():
                 entity=template,
                 action='export',
                 action_condition=lambda t: export_template is None,
+                template=search_entity(),
```
Please remove, I think this isn't really needed.

maorlipchuk replied:
yes, it was added by mistake, thanks!

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -326,32 +352,33 @@ def main():
             if template is None:
-                # Test if template is registered:
-                template = templates_module.search_entity()
+                template = search_entity()
```
Why this change is needed?

maorlipchuk replied:
It was by mistake, I will change it back

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -326,32 +352,33 @@ def main():
-                # Register the template into the system:
                 template_service.register(
+                    allow_partial_import=module.params['allow_partial_import']
+                    if module.params['allow_partial_import'] else None,
```
No need for `if` here, because the default is `None` anyway and if `False`, `None` is passed, which is wrong.

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -41,17 +41,26 @@
     id:
         description:
             - "ID of the Virtual Machine to manage."
+        version_added: "2.4"
```
Please remove.

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -41,17 +41,26 @@
         description:
-            - "Should the Virtual Machine be running/stopped/present/absent/suspended/next_run."
+            - "Should the Virtual Machine be running/stopped/present/absent/suspended/next_run/registered.
+               When C(state) is R(registered) and the unregistered VM's name
```
I(registered)

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -354,6 +363,28 @@
+    cluster: mycluster
+    name: myvm
+
+# Register VM
```
# Register VM using id

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1243,6 +1289,49 @@ def main():
+                vm_service = vms_service.vm_service(vm.id)
+                vm_service.register(
+                    allow_partial_import=module.params['allow_partial_import']
+                    if module.params['allow_partial_import'] else None,
```
please remove the `if`

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1243,6 +1289,49 @@ def main():
+            )
+            changed = False
+            if vm is None:
+                vm = search_entity()
```
search_entity -> vms_module.search_entity?

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1054,15 +1086,28 @@ def control_state(vm, vms_service, module):
+        timeout = module.params['timeout']
+        poll_interval = module.params['poll_interval']
+        while time.time() < start + timeout:
+            vm = search_entity()
```
vms_module.search_entity?

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -1054,15 +1086,28 @@ def control_state(vm, vms_service, module):
             )
 
 
+def wait_for_import(module, vms_module, vms_service):
```
You can consider add this function to `ovirt.py` as it's already used twice.

maorlipchuk replied:
will do

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -103,6 +104,13 @@
+        description:
+          - "If I(iscsidiscover) it means that the iscsi attribute is being
+             used to discover targets"
+          - "If I(iscsilogin) it means that the iscsi attribute is being
```
If C(state) is I(iscsilogin)

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -103,6 +104,13 @@
                Availability cluster."
+    iscsi:
+        description:
+          - "If I(iscsidiscover) it means that the iscsi attribute is being
```
Please change to, just to make it clear :)
```
If C(state) is I(iscsilogin)
```

maorlipchuk replied:
done, for iscsidiscover and login

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -336,10 +372,15 @@ def main():
         supports_check_mode=True,
+        required_if=[
+            ['state', 'iscsi_login', ['iscsi']],
+            ['state', 'iscsi_discover', ['iscsi']]
```
s/iscsi_login/iscsilogin
s/iscsi_discover/iscsidiscover

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -386,6 +427,34 @@ def main():
+            iscsi_targets = hosts_service.service(host_id).iscsi_discover(
+                iscsi=otypes.IscsiDetails(
+                    port=module.params['iscsi']['port']
+                    if module.params['iscsi']['port'] else None,
```
this `if` isn't needed , no?

maorlipchuk replied:
I haven't verified discover yet, I will remove it and verify

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_hosts.py:
```diff
@@ -386,6 +427,34 @@ def main():
+            ret = hosts_module.action(
+                action='iscsi_login',
+                iscsi=otypes.IscsiDetails(
+                    port=module.params['iscsi']['port'] if module.params['iscsi']['port'] else None,
```
This `if` isn't needed , no ?

maorlipchuk replied:
removed., and will update once I will verify as well

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_templates.py:
```diff
@@ -57,6 +60,10 @@
             - "Name of the cluster, where template should be created/imported."
+    allow_partial_import:
+        description:
+            - "Boolean indication whether to allow partial registeration of a template when C(state) is registered."
```
s/registeration/registration

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -522,7 +552,7 @@
     type: dict
 '''
-
+import time
```
no need for this import now?

maorlipchuk replied:
nope, removed

Commit:
ovirt_hosts: Add iscsilogin to ovirt hosts.

Add functionality of iscsi login to ovirt hosts to be used to connect to
iscsi targets and to be able to import iSCSI storage domain eventually.

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,123 @@
+# Gather facts about all Templates which relate to a storage domain and
+# are unregistered:
+- ovirt_storage_templates_facts:
+    pattern: unregistered=True
```
Please remove it also from here, use just `unregistered: True`

maorlipchuk replied:
sure, done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,123 @@
+        templates_service = storage_domain_service.templates_service()
+
+        # Find the the unregistered Template we want to register:
+        if (module.params['pattern'] == "unregistered=True"):
```
And from here.

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_vms_facts.py:
```diff
@@ -0,0 +1,122 @@
+# Gather facts about all VMs which relate to a storage domain and
+# are unregistered:
+- ovirt_vms_facts:
+    pattern: unregistered=True
```
Same for VMs.

maorlipchuk replied:
done

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,123 @@
+
+def main():
+    argument_spec = ovirt_facts_full_argument_spec(
+        pattern=dict(default='', required=False),
```
last thing hopefully :)) please remove the pattern here...

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_templates_facts.py:
```diff
@@ -0,0 +1,123 @@
+        if module.params.get('unregistered'):
+            templates = templates_service.list(unregistered=True)
+        else:
+            templates = templates_service.list(module.params['pattern'])
```
and here ,,,

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_vms_facts.py:
```diff
@@ -0,0 +1,122 @@
+
+def main():
+    argument_spec = ovirt_facts_full_argument_spec(
+        pattern=dict(default='', required=False),
```
and here ,,,

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_storage_vms_facts.py:
```diff
@@ -0,0 +1,122 @@
+        if module.params.get('unregistered'):
+            vms = vms_service.list(unregistered=True)
+        else:
+            vms = vms_service.list(module.params['pattern'])
```
and here ,,,

machacekondra reviewed (commented):
Last comment, then it's ready to be shipped 👍

machacekondra commented:
@mwperina @ryansb shipit

machacekondra reviewed (approved):
shipit

machacekondra commented:
@maorlipchuk  Can you please rebase and resolve conficlits?

mwperina reviewed (approved):
Looks good to me, thanks! 👍

mwperina commented:
shipit
